### ISSUE TITLE
Some inconsistencies with generic cosmetic disable

### ISSUE THREAD

**Issue by joey04** [uBlockOrigin/uBlock-issues#233]
### Prerequisites

- [ ] I verified that this is not a filter issue
    - Filter issues MUST be reported at [filter issue tracker](https://github.com/uBlockOrigin/uAssets/issues)
    
- [ ] This is not a support issue or a question
    - Support issues and questions are handled at [/r/uBlockOrigin](https://old.reddit.com/r/uBlockOrigin/)
    
- [x] I performed a cursory search of the issue tracker to avoid opening a duplicate issue
    - Your issue may already be reported.
- I tried to reproduce the issue when...
    - [ ] uBlock Origin is the only extension
    - [ ] uBlock Origin with default lists/settings
    - [ ] using a new, unmodified browser profile
- [x] I am running the latest version of uBlock Origin
- [ ] I checked the [documentation](https://github.com/gorhill/uBlock/wiki) to understand that the issue I report is not a normal behavior

### Description

There are 3 ways to disable generic cosmetics:
1. Full cosmetic disable switch
2. $generichide
3. ignoreGenericCosmeticFilters option

All 3 do their main job of disabling generics, but there are inconsitencies with other concerns:
* The first two will lead to the content script fully disabling domSurveyor, but the third doesn't. 
* Only $generichide factors into in the recent fix for https://github.com/gorhill/uBlock/issues/2763.

I fixed these for myself by integrating ignoreGenericCosmeticFilters into both PageStore.init() and reverselookup ignoreGeneric.

**Comment by gorhill** [comment]
> The first two will lead to the content script fully disabling domSurveyor, but the third doesn't

There are no steps to reproduce so I had to come up with mines:

1. Go to _"Filter lists"_ pane
1. Be sure that _"Ignore generic cosmetic filters"_ is not check -- uncheck and _"Apply changes"_ otherwise.
1. Open `https://news.ycombinator.com/` in a new tab
1. Open dev tools for the page (ctrl-shift-i)
1. Bring up uBO's `contentscript.js` and set a breakpoint for [the line where the surveyor is started](https://github.com/gorhill/uBlock/blob/3c85c0319462ca331d53c350fba4bc6c1b2ef96f/src/js/contentscript.js#L1393)
1. Force a reload of the tab
1. Confirm that the break point is taken
1. Check _"Ignore generic cosmetic filters"_ and click _"Apply changes"_
1. Force a reload of the tab
1. Observe that the breakpoint is not taken

The DOM surveyor is not launched because the response from the main uBO process says that `noDOMSurveying` is `true`. `noDOMSurveying` is true because `µBlock.cosmeticFilteringEngine.hasGenericHide` is `false`.

**Comment by joey04** [comment]
Interesting. I was reporting this purely after a code readthrough and removal of the no-cosmetic switch (which I've never used).

I didn't realize that `noDOMSurveying` would prevent its launching, given that I didn't read through the byzantine CFE routine. (Which I very much appreciate for the great work it does but would take lots of time to come up to speed on.)

The second point still stands. It's an easy fix by adding `µBlock.userSettings.ignoreGenericCosmeticFilters ||` to [this line](https://github.com/gorhill/uBlock/blob/master/src/js/reverselookup.js#L186).

**Comment by gorhill** [comment]
I still need to fix the second point, which causes similar issue as the one reported in <https://github.com/gorhill/uBlock/issues/2763>. However I am evaluating whether uBO should even compile the generic cosmetic filters. Not coompiling would solve the issue here _and_ reduce the size of the compiled list -- a good thing for Firefox for Android (where generic cosmetic filters are not used by default). This fix would be a bit more complicated though, so I have not yet decided what I will do.

**Comment by joey04** [comment]
> I am evaluating whether uBO should even compile the generic cosmetic filters.

Personally I like the option as-is. It's cool to see the count of generics that were compiled but not loaded (about 20k in EasyList). Compiling them also has the advantage of being a seamless transition when the option is toggled on/off then click 'apply'.

And I can speak to the other side, because in my fork I added the ability to only compile a specific subset of the lists. I made it a check-box option exactly the same as ignoreGenerics because I don't want it for all of my browsers. It's definitely not seamless to transition when toggling on/off. You'd probably have to purge caches and recompile all lists to get a full transition. For me, this is a don't care becuase it's just for myself and I never change it for each of my browsers. But as a general uBO feature it's a dicey prospect, unless you were to recompile everything after the option is toggled and click 'apply'.
